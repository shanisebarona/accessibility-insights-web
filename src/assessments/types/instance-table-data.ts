// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { GeneratedAssessmentInstance } from 'common/types/store-data/assessment-result-data';

export type InstanceTableHeaderType = 'none' | 'default';

export interface InstanceTableRow<P = {}> {
    key: string;
    statusChoiceGroup: JSX.Element;
    visualizationButton?: JSX.Element;
    instance: GeneratedAssessmentInstance<P>;
}

export interface InstanceTableColumn {
    key: string;
    name: string;
    onRender: (item: InstanceTableRow) => JSX.Element;
}
