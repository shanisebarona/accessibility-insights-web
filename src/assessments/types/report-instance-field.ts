// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import {
    ColumnValue,
    ColumnValueBag,
    isScalarColumnValue,
} from 'common/types/property-bag/column-value-bag';
import { PropertyBagColumnRendererConfig } from 'common/types/property-bag/property-bag-column-renderer-config';
import { TestStepInstance } from 'common/types/store-data/assessment-result-data';

export type ReportInstanceField = {
    key: string;
    label: string;
    getValue: (instance: Partial<TestStepInstance>) => ColumnValue;
};
export type ReportInstanceFieldMap = { [KEY in string]?: ReportInstanceField };

export type ReportInstanceFields = ReportInstanceField[];

type HasPropertyBag<PB> = { propertyBag: PB };
type PropertyBagKey<PB> = PB[keyof PB];

function fromPropertyBagField<PB>(label: string, key: keyof PB & string): ReportInstanceField {
    function getValue(i: HasPropertyBag<PB>): string {
        return i.propertyBag && i.propertyBag[key] && (i.propertyBag[key] as any).toString();
    }
    return { key, label, getValue };
}

function fromColumnValueBagField<PB extends ColumnValueBag>(
    label: string,
    key: keyof PB & string,
    defaultValue = '-',
): ReportInstanceField {
    function getValue(i: HasPropertyBag<PB>): string | PropertyBagKey<PB> {
        const value = i.propertyBag && i.propertyBag[key];
        return isValid(value) ? value : defaultValue;
    }

    return { key, label, getValue };
}

function fromPropertyBagFunction<PB>(
    label: string,
    key: string,
    accessor: (bag: PB) => string,
): ReportInstanceField {
    function getValue(i: HasPropertyBag<PB>): string {
        return i.propertyBag && accessor(i.propertyBag);
    }
    return { key, label, getValue };
}

const common: ReportInstanceFieldMap = {
    comment: { key: 'comment', label: 'Comment', getValue: i => i.description },
    snippet: { key: 'snippet', label: 'Snippet', getValue: i => i.html },
    path: { key: 'path', label: 'Path', getValue: i => i.target && i.target.join(', ') },
    manualSnippet: { key: 'manualSnippet', label: 'Code Snippet', getValue: i => i.html },
    manualPath: { key: 'manualPath', label: 'CSS Selector', getValue: i => i.selector },
};

function isValid(value: ColumnValue): ColumnValue {
    if (!value) {
        return false;
    }
    if (!isScalarColumnValue(value) && Object.keys(value).length === 0) {
        return false;
    }
    return true;
}

function fromColumns<T extends ColumnValueBag>(
    cfg: PropertyBagColumnRendererConfig<T>[],
): ReportInstanceField[] {
    return cfg.map(fromColumnConfig);

    function fromColumnConfig({
        propertyName,
        defaultValue,
        displayName,
    }: PropertyBagColumnRendererConfig<T>): ReportInstanceField {
        const getValue = (inst: HasPropertyBag<T>) => {
            const value = inst.propertyBag && inst.propertyBag[propertyName];
            return isValid(value) ? value : defaultValue;
        };

        return {
            key: propertyName,
            label: displayName,
            getValue,
        };
    }
}

function fromSnippet(key: string, label: string): ReportInstanceField {
    return { key, label, getValue: i => i.html };
}

export const ReportInstanceField = {
    fromPropertyBagField,
    fromColumnValueBagField,
    fromPropertyBagFunction,
    fromColumns,
    fromSnippet,
    common,
};
