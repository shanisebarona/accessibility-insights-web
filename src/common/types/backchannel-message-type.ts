// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

export type BackchannelStoreRequestMessage = {
    messageId: string; // same as WindowMessage
    messageType: 'backchannel_window_message.store_request';
    stringifiedMessageData: string; // JSON.stringify(originalMessage)
};

export type BackchannelRetrieveRequestMessage = {
    messageId: string; // same as WindowMessage
    messageType: 'backchannel_window_message.retrieve_request';
};

export type BackchannelRetrieveResponseMessage = {
    messageId: string; // same as WindowMessage
    messageType: 'backchannel_window_message.retrieve_response';
    stringifiedMessageData: string;
};
