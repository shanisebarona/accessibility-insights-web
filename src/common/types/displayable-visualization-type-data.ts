// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
export interface DisplayableVisualizationTypeData {
    title: string;
    subtitle?: JSX.Element;
    enableMessage: string;
    toggleLabel: string;
    linkToDetailsViewText: string;
}
