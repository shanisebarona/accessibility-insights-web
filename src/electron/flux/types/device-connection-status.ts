// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
export enum DeviceConnectionStatus {
    Unknown,
    Connected,
    Disconnected,
}
