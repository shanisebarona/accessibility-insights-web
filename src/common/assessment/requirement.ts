// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { AssessmentsProvider } from 'assessments/types/assessments-provider';
import { Requirement } from 'assessments/types/requirement';
import { keys, sortBy } from 'lodash';
import { ManualTestStatusData, TestStepData } from '../types/manual-test-status';
import { VisualizationType } from '../types/visualization-type';

export type RequirementDefinition = Requirement;
export type RequirementData = TestStepData;

export type RequirementResult = {
    definition: RequirementDefinition;
    data: RequirementData;
};
export type RequirementOrdering = RequirementOrderPart | RequirementOrderPart[];
export type RequirementOrderPart = (result: RequirementResult) => string | number;

export function getRequirementsResults(
    provider: AssessmentsProvider,
    visualizationType: VisualizationType,
    stepStatus: ManualTestStatusData,
): RequirementResult[] {
    const test = provider.forType(visualizationType);

    function result(key): RequirementResult {
        const definition = provider.getStep(visualizationType, key);
        const data = stepStatus[key];
        return { definition, data };
    }

    const unsorted = keys(stepStatus).map(result);

    return sortBy(unsorted, test.requirementOrder);
}
