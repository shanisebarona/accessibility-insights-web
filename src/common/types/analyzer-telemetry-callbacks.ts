// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { AxeAnalyzerResult } from 'common/types/axe-analyzer-result';
import { VisualizationType } from 'common/types/visualization-type';
import {
    RuleAnalyzerScanTelemetryData,
    IssuesAnalyzerScanTelemetryData,
    NeedsReviewAnalyzerScanTelemetryData,
} from '../extension-telemetry-events';

export type IAnalyzerTelemetryCallback = ForRuleAnalyzerScanCallback;

export type ForRuleAnalyzerScanCallback = (
    analyzerResult: AxeAnalyzerResult,
    scanDuration: number,
    elementsScanned: number,
    testVisualizationType: VisualizationType,
    requirementName?: string,
) => RuleAnalyzerScanTelemetryData;

export type ForIssuesAnalyzerScanCallback = (
    analyzerResult: AxeAnalyzerResult,
    scanDuration: number,
    elementsScanned: number,
    testVisualizationType: VisualizationType,
) => IssuesAnalyzerScanTelemetryData;

export type ForNeedsReviewAnalyzerScanCallback = (
    analyzerResult: AxeAnalyzerResult,
    scanDuration: number,
    elementsScanned: number,
    testVisualizationType: VisualizationType,
) => NeedsReviewAnalyzerScanTelemetryData;
