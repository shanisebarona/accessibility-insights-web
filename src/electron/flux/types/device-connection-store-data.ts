// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { DeviceConnectionStatus } from 'electron/flux/types/device-connection-status';

export interface DeviceConnectionStoreData {
    status: DeviceConnectionStatus;
}
