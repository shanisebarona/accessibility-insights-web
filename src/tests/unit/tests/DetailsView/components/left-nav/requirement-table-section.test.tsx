// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { AssessmentsProvider } from 'assessments/types/assessments-provider';
import { Requirement } from 'assessments/types/requirement';
import { ManualTestStepResult } from 'common/types/store-data/assessment-result-data';
import { FeatureFlagStoreData } from 'common/types/store-data/feature-flag-store-data';
import { PathSnippetStoreData } from 'common/types/store-data/path-snippet-store-data';
import {
    RequirementTableSection,
    RequirementTableSectionProps,
} from 'DetailsView/components/left-nav/requirement-table-section';
import { AssessmentInstanceTableHandler } from 'DetailsView/handlers/assessment-instance-table-handler';
import { shallow } from 'enzyme';
import * as React from 'react';
import { DictionaryStringTo } from 'types/common-types';

describe('RequirementTableSection', () => {
    let props: RequirementTableSectionProps;
    let manualRequirementResultMapStub: DictionaryStringTo<ManualTestStepResult>;
    let assessmentInstanceTableHandlerStub: AssessmentInstanceTableHandler;
    let assessmentProviderStub: AssessmentsProvider;
    let featureFlagStoreDataStub: FeatureFlagStoreData;
    let pathSnippetStoreDataStub: PathSnippetStoreData;

    beforeEach(() => {
        assessmentInstanceTableHandlerStub = {
            changeRequirementStatus: null,
        } as AssessmentInstanceTableHandler;
        manualRequirementResultMapStub = {
            'some manual test step result id': null,
        };
        assessmentProviderStub = {
            all: null,
        } as AssessmentsProvider;
        featureFlagStoreDataStub = {
            'some feature flag': true,
        };
        pathSnippetStoreDataStub = {
            path: null,
        } as PathSnippetStoreData;

        props = {
            assessmentNavState: {
                selectedTestSubview: 'some test view',
                selectedTestType: -1,
            },
            manualRequirementResultMap: manualRequirementResultMapStub,
            assessmentInstanceTableHandler: assessmentInstanceTableHandlerStub,
            assessmentsProvider: assessmentProviderStub,
            featureFlagStoreData: featureFlagStoreDataStub,
            pathSnippetStoreData: pathSnippetStoreDataStub,
            requirement: {
                isManual: false,
                getDefaultMessage: _ => null,
                instanceTableHeaderType: 'none',
            },
        } as RequirementTableSectionProps;
    });

    test('render ManualTestStepView when requirement is manual', () => {
        props.requirement = {
            isManual: true,
        } as Requirement;

        const testObject = shallow(<RequirementTableSection {...props} />);

        expect(testObject.getElement()).toMatchSnapshot();
    });

    test('render Spinner when scanning in progress', () => {
        props.scanningInProgress = true;

        const testObject = shallow(<RequirementTableSection {...props} />);

        expect(testObject.getElement()).toMatchSnapshot();
    });

    test('render instance table', () => {
        const testObject = shallow(<RequirementTableSection {...props} />);

        expect(testObject.getElement()).toMatchSnapshot();
    });
});
