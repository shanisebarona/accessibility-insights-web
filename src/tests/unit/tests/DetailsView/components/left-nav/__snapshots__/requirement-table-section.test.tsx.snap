// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequirementTableSection render ManualTestStepView when requirement is manual 1`] = `
<ManualTestStepView
  assessmentInstanceTableHandler={
    Object {
      "changeRequirementStatus": null,
    }
  }
  assessmentsProvider={
    Object {
      "all": null,
    }
  }
  featureFlagStoreData={
    Object {
      "some feature flag": true,
    }
  }
  manualTestStepResultMap={
    Object {
      "some manual test step result id": null,
    }
  }
  pathSnippetStoreData={
    Object {
      "path": null,
    }
  }
  step="some test view"
  test={-1}
/>
`;

exports[`RequirementTableSection render Spinner when scanning in progress 1`] = `
<StyledSpinnerBase
  className="details-view-spinner"
  label="Scanning"
  size={3}
/>
`;

exports[`RequirementTableSection render instance table 1`] = `
<React.Fragment>
  <h3>
    Instances
  </h3>
  <AssessmentInstanceTable
    assessmentInstanceTableHandler={
      Object {
        "changeRequirementStatus": null,
      }
    }
    assessmentNavState={
      Object {
        "selectedTestSubview": "some test view",
        "selectedTestType": -1,
      }
    }
    getDefaultMessage={[Function]}
    instanceTableHeaderType="none"
  />
</React.Fragment>
`;
